### PR TITLE
Use context managers to close files properly and fix tests on PyPy

### PR DESCRIPTION
**Pull request by mgorny**
**What kind of change does this PR introduce?**
  - [x] bug fix
  - [ ] feature
  - [ ] docs update
  - [x] tests/coverage improvement
  - [x] refactoring
  - [ ] other

**What is the related issue number (starting with `#`)**

#1308 (partially related, it's an old bug filed for PyPy2)

**What is the current behavior?** (You can also link to an open issue here)

The code prior to this change used four approaches for closing files:

1. Using a context manager (`with` clause).
2. Using a try/finally clause.
3. Closing the file in the same scope (unreliable: file object can leak on exception).
4. Not closing open files at all.
    
The last point is a real problem for PyPy since it does not GC unreachable objects as aggressively as CPython does.  While leaving a function scope on CPython causes the file objects private to it to be destroyed (and therefore closed), in PyPy they can stay dangling
for some time.  When combines with buffered writes, this means that writes can still remain pending after returning from function.

**What is the new behavior (if this is a feature change)?**

Context managers are used consistently everywhere. This gives me 100% test pass with PyPy3.7.

**Other information**:

**Checklist**:

  - [x] I think the code is well written
  - [x] I wrote [good commit messages][1]
  - [x] I have [squashed related commits together][2] after the changes have been approved
  - [x] Unit tests for the changes exist
  - [x] Integration tests for the changes exist (if applicable)
  - [x] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [x] Documentation reflects the changes (-- I don't think any doc changes are necessary)
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

### CODE REVIEW

**Comment by mgorny** [comment]
It seems that the AppVeyor failure is a flaky test.

**Comment by graingert** [comment]
Can this warning ignore be removed now? https://github.com/cherrypy/cherrypy/blob/main/pytest.ini#L41

File: cherrypy/lib/covercp.py [hunk 1]
```diff
@@ -334,26 +334,26 @@ def menu(self, base='/', pct='50', showpct='',
...
+                elif lineno in statements:
+                    template = TEMPLATE_LOC_COVERED
+                else:
+                    empty_the_buffer = False
+                    buffer.append((lineno, line))
+                if empty_the_buffer:
+                    for lno, pastline in buffer:
+                        yield template % (lno, cgi.escape(pastline))
+                    buffer = []
+                    yield template % (lineno, cgi.escape(line))
```
Comment (graingert) [inline cherrypy/lib/covercp.py]: because there's a `yield` inside this `with ` you're relying on the GC to `.close()` this generator to exit the `with` block using a `GeneratorExit`

I think you can fix it by using:

```python
yield from self.annotated_file(filename, statements, excluded,
                                        missing):
```

in `def report(self, name):`

Comment (mgorny) [reply cherrypy/lib/covercp.py]: Ok, I've changed that. Alternatively, I suppose we could just put the list from `readlines()` into a variable and have it closed before the iteration.

Comment (graingert) [reply cherrypy/lib/covercp.py]: :facepalm: oh I only just noticed the `readlines()` call.

```python
with open(filename, 'r') as f:
    lines = f.readlines()
```

would be much more clear that the file is immediately consumed and closed.

however maybe that's a mistake and the file was intended to be streamed, and the code should be:

```python
for lineno, line in enumerate(source):
```

**Comment by mgorny** [comment]
> Can this warning ignore be removed now? https://github.com/cherrypy/cherrypy/blob/main/pytest.ini#L41

I'm afraid not. There are more leaking files but the traceback is entirely unhelpful to me. I'm not even sure if they're bugs in cherrypy or just the test suite taking shortcuts somewhere.

File: cherrypy/lib/covercp.py (hunk 1 above)
Comment (graingert) [reply cherrypy/lib/covercp.py]: Originally the file was streamed in this generator and then it was changed to `.readlines()` https://github.com/cherrypy/cherrypy/commit/596b1f603047a407bf8949b018671c7de13e9f2f#diff-342872edf8db6cab0554b70351960b4850b8eb7120003bad9e2d074c69a7c758L335-R333

Comment (mgorny) [reply cherrypy/lib/covercp.py]: Ok, I'll do the first snippet but I suppose there's no harm in leaving the `yield from` change as well.

**Comment by webknjaz** [comment]
> I'm not even sure if they're bugs in cherrypy or just the test suite taking shortcuts somewhere.

I think this may be both. There were some corner cases I've been unable to debug. Ideally, we shouldn't have any ignores like that, of course. The only reason it is there is that some of them are hard to track down, especially, when they are flaky.

**Comment by mgorny** [comment]
Just to be clear, I'm only @Gentoo packager for Python stuff and I don't even really know how to use CherryPy (nor wanna learn, sorry), let alone figure out how it deals with keeping resources like open files.

File: cherrypy/lib/covercp.py [hunk 2]
```diff
@@ -364,9 +364,8 @@ def report(self, name):
                                        fullpath=name,
                                        pc=pc)
         yield '<table>\n'
-        for line in self.annotated_file(filename, statements, excluded,
```
Comment (webknjaz) [inline cherrypy/lib/covercp.py]: Was this leaking file descriptors? I don't fully understand why it is in the scope of this PR (I'd appreciate if it was as atomic as possible).

Comment (mgorny) [reply cherrypy/lib/covercp.py]: No, we thought it was originally necessary because of changes in `annotated_file()` but it's not now. I'll revert it then.

Comment (graingert) [reply cherrypy/lib/covercp.py]: @webknjaz it was leaking an unclosed generator (`annotated_file()`), which then went on to leak a file descriptor - the `yield` has been moved outside of the `with open(filename) as source): ` block and so now only the generator is leaked

**Comment by webknjaz** [comment]
@mgorny I know and I highly appreciate your contribution! Unfortunately, the current CI setup is also in a bad state so keep in mind that the AppVeyor tests may be failing for unrelated reasons.

P.S. Note that our release automation got broken over time so I won't be able to release it right away. So I suppose you may need to integrate this patch into the ebuild via https://github.com/cherrypy/cherrypy/pull/1946.patch.

**Review by webknjaz — APPROVED** [review]
Brilliant! Thanks for the patch! I'll wait a bit for the Circle CI status to be reported and will merge then.

**Comment by mgorny** [comment]
Don't worry about the release. Right now my biggest blocker is still the horror called twisted.
